### PR TITLE
#1111 New CustomerInvoicePage filtering feature

### LINKED ISSUE #1111: New CustomerInvoicePage filtering feature
## Is your feature request related to a problem? Please describe.

The new CustomerInvoicePage needs a filtering feature.

## Describe the solution you'd like

The same as sorting - need an array of realm objects to display data, but also have access to a realm results backing data.

options:
- Filter on the backing data using realm queries, apply sorting, convert to an array. I think clearing sorting and simply filtering and converting is a viable option. Is filtering ever needed on getter fields?

OR

- Filter and return the data array using basic array methods

I'd suspect that realm would be a fair bit quicker for this. Especially if we index common filter fields, which is a valid use case. 

Should filtering be done on potentially many fields?

## Describe alternatives you've considered

As above, couple options.

## Additional context

Might need to check the performance of both options, highly suspect realm queries to be faster. Is clearing the current sort state terrible after a search?

Should also debounce this

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1111

**Forgot to rename this branch before publishing, oopsies**
## Change summary

- Adds a simple filtering mechanism
- Went with the simple approach, using realm to query a `backingData` in state, returning a filtered version of that.
- Could add realm sorting on top, but would also have to determine if the field is a getter/setter etc etc - just doing it simple

## Testing
- [ ] `CustomerInvoicePage` filtering functionality works

### Related areas to think about

As above - dealing with sorting while filtering. I think keeping this simple for now is a good option, personally - just removing the sorting. Hard to know if the added complexity is really actually worth it. I don't see adding a filter then pushing a sort button to be odd

### CODE REVIEW

File: src/pages/CustomerInvoicePage.js [hunk 1]
```diff
@@ -390,7 +406,12 @@ export const CustomerInvoicePage = ({
     });
   };
 
-  const onSearchChange = () => {};
+  const onSearchChange = searchTerm => {
+    dispatch(filterData(searchTerm));
+  };
+
+  const instantDebouncedDispatch = useMemo(() => debounce(dispatch, 250, true), []);
```
Comment (Chris-Petty) [inline src/pages/CustomerInvoicePage.js]: unused?

File: src/pages/CustomerInvoicePage.js [hunk 2]
```diff
@@ -390,7 +406,12 @@ export const CustomerInvoicePage = ({
     });
   };
 
-  const onSearchChange = () => {};
+  const onSearchChange = searchTerm => {
+    dispatch(filterData(searchTerm));
```
Comment (Chris-Petty) [inline src/pages/CustomerInvoicePage.js]: would be cleaner and make the func given to onChange have this better variable name if you just put debounce in here?

File: src/pages/CustomerInvoicePage.js (hunk 1 above)
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: Used for a header cell

File: src/pages/CustomerInvoicePage.js (hunk 2 above)
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: I think I should do a PR for `SearchBar` to accept an optional `action` prop so we can bypass this function altogether. I haven't tried but I *think* putting `debounce` in here will return a new function each call, resulting in all invocations not having the same reference to the timeout variable to cancel consecutive calls?

**Comment by joshxg** [comment]
Have added post filtering sorting : 25538b38b4bb68f42afc6496054e666897f4569f

(Just duplicated the `columnKeyToDataType` mapping object until we find a nice home

**Pull request merged**